language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToConstantCodeAction.java

```diff
@@ -0,0 +1,119 @@
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.CodeActionKind;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.TextEdit;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Code Action for extracting to a constant.
+ *
+ * @since 2201.2.0
+ */
+@JavaSPIService("org.ballerinalang.langserver.commons.codeaction.spi.LSCodeActionProvider")
+public class ExtractToConstantCodeAction implements RangeBasedCodeActionProvider {
+
+    public static final String NAME = "Extract To Constant";
+    private static final String CONSTANT_NAME_PREFIX = "CONSTANT";
+
+    public List<SyntaxKind> getSyntaxKinds() {
+        return List.of(SyntaxKind.BOOLEAN_LITERAL, SyntaxKind.NUMERIC_LITERAL,
+                SyntaxKind.STRING_LITERAL, SyntaxKind.BINARY_EXPRESSION);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<CodeAction> getCodeActions(CodeActionContext context,
+                                                    RangeBasedPositionDetails posDetails) {
+        
+        if (context.currentSyntaxTree().isEmpty() || context.currentSemanticModel().isEmpty()) {
+            return Collections.emptyList();
+        }
+        Node node = posDetails.matchedCodeActionNode();
+        if (node.parent().kind() == SyntaxKind.CONST_DECLARATION ||
+                node.parent().kind() == SyntaxKind.INVALID_EXPRESSION_STATEMENT ||
+                (node.kind() != SyntaxKind.NUMERIC_LITERAL && node.kind() != SyntaxKind.STRING_LITERAL && 
+                        node.kind() != SyntaxKind.BOOLEAN_LITERAL && node.kind() != SyntaxKind.BINARY_EXPRESSION)) {
+            return Collections.emptyList();
+        }
+        
+        // If the node is a BinaryExpressionNode, check whether it contains only BasicLiteralNodes
+        if (node.kind() == SyntaxKind.BINARY_EXPRESSION) {
+            ConstantVisitor constantVisitor = new ConstantVisitor();
+            node.accept(constantVisitor);
+            if (constantVisitor.getInvalidNode()) {
+                return Collections.emptyList();
+            }
+        }
+        
+        Set<String> visibleSymbolNames = context.visibleSymbols(context.cursorPosition()).stream()
+                .map(Symbol::getName)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(Collectors.toSet());
+        String constName = NameUtil.generateTypeName(CONSTANT_NAME_PREFIX, visibleSymbolNames);
+
+        String value = node.toSourceCode().strip();
+        LineRange replaceRange = node.lineRange();
+        Optional<TypeSymbol> typeSymbol = context.currentSemanticModel().get().typeOf(node);
+        if (typeSymbol.isEmpty()) {
+            return Collections.emptyList();
+        }
+        
+        Node modPartNode = node;
+        while (modPartNode.parent().kind() != SyntaxKind.MODULE_PART) {
```

Comment:
Why can't we simply do `context.currentSyntaxTree().get().rootNode()` to get the module part node?

Reply:
Used `context.currentSyntaxTree().get().rootNode()` instead.